internal/connection/connection.go

```diff
@@ -29,23 +31,60 @@ import (
 	"google.golang.org/grpc/credentials/insecure"
 )
 
+const (
+	// The duration after which a gRPC connection is closed due to inactivity
+	IDLE_TIMEOUT = time.Minute * 5
+)
+
 // Connection struct consists of to NodeID, DriverName, Capabilities for the controller
 // to pick sidecar connection and grpc Client to connect to the sidecar.
 type Connection struct {
+	sync.Mutex
+
 	Client       *grpc.ClientConn
 	Capabilities []*identity.Capability
 	Namespace    string
 	Name         string
 	NodeID       string
 	DriverName   string
 	Timeout      time.Duration
+
+	// Holds the internal state of the connection
+	connected  bool
+	enableAuth bool
+	endpoint   string
+	nodeID     string
+	driverName string
+	podName    string
+
+	// Used to cancel any existing timers in case of re-connects
+	cancelIdle context.CancelFunc
+
+	lastAccessTime atomic.Int64
 }
 
-// NewConnection establishes connection with sidecar, fetches capability and returns Connection object
-// filled with required information.
-func NewConnection(ctx context.Context, endpoint, nodeID, driverName, namespace, podName string, enableAuth bool) (*Connection, error) {
-	var opts []grpc.DialOption
-	if enableAuth {
+// accessTimeInterceptor is an unary interceptor which updates the lastAccessTime
+// on `Connection` struct to `time.Now()` on each RPC call
+func accessTimeInterceptor(conn *Connection) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		conn.lastAccessTime.Store(time.Now().UnixNano())
+
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+}
+
+func (c *Connection) Connect(ctx context.Context) error {
```

Comment:
is the ctx parameter really useful (is it used at all?), maybe it should be allocated within this `Connect()` function instead

Reply:
Done